agents/dqn.py

```diff
@@ -1,24 +1,132 @@
+
 from agents import TorchAgent
+from ml.dqn_encoders import BaseQNetwork
 from ml.encoders import BaseEncoder
+from ml.utils import TensorWithMask
+from utils.bag_trajectory import BaseBagTrajectoryMemory
 
 
-# TODO[Vladimir Baikalov]: Implement
-class DQNAgent(TorchAgent, config_name='dqn'):
+def _with_random_research(
+        argmax_next_neighbor: Tensor,
+        neighbor: TensorWithMask,
+        prob: float
+):
+    mask = torch.bernoulli(torch.full(argmax_next_neighbor.shape, prob)).int()
+    reverse_mask = torch.ones(argmax_next_neighbor.shape).int() - mask
+
+    uniform_weights = neighbor.mask
+    uniform_distr = Categorical(probs=uniform_weights / uniform_weights.sum(dim=1))
+    random_next_neighbor = neighbor.padded_values[torch.unsqueeze(uniform_distr.sample(), dim=1)]
 
-    def __init__(self, parts):
+    return argmax_next_neighbor * mask + random_next_neighbor * reverse_mask
+
+
+class DQNAgent(TorchAgent, config_name='dqn'):
+    def __init__(
+            self,
+            current_node_idx_prefix: str,
+            destination_node_idx_prefix: str,
+            neighbors_node_ids_prefix: str,
+            node_idx: int,
+            q_network: BaseQNetwork,
+            discount_factor: float,
+            research_prob: float,
+            bag_trajectory_memory: BaseBagTrajectoryMemory,
+            sample_size: int,
+    ):
+        assert 0 < discount_factor < 1, 'Incorrect `discount_factor` choice'
+        assert 0 < research_prob < 1, 'Incorrect `discount_factor` choice'
         super().__init__()
-        self._parts = parts
+
+        self._current_node_idx_prefix = current_node_idx_prefix
+        self._destination_node_idx_prefix = destination_node_idx_prefix
+        self._neighbors_node_ids_prefix = neighbors_node_ids_prefix
+
+        self._node_idx = node_idx
+        self._q_network = q_network
+        self._discount_factor = discount_factor
+        self._research_prob = research_prob
+        self._bag_trajectory_memory = bag_trajectory_memory
+        self._sample_size = sample_size
 
     @classmethod
     def create_from_config(cls, config):
         return cls(
-            parts=[
-                BaseEncoder.create_from_config(part)
-                for part in config['parts']
-            ]
+            current_node_idx_prefix=config['current_node_idx_prefix'],
+            destination_node_idx_prefix=config['destination_node_idx_prefix'],
+            neighbors_node_ids_prefix=config['neighbors_node_ids_prefix'],
+            q_network=BaseEncoder.create_from_config(config['q_network']),
+            discount_factor=config.get('discount_factor', 0.99),
+            research_prob=config.get('research_prob', 0.1),
+            bag_trajectory_memory=BaseBagTrajectoryMemory.create_from_config(config['path_memory'])
+            if 'path_memory' in config else None,
+            node_idx=config.get('node_idx', None),
+            sample_size=config.get('sample_size', 100)
```

Comment:
This is not consistent with the `__init__` function signature. If you are using `config.get(...)` in this function, then in the `__init__` the same variable will also have a default value, otherwise, you do it implicitly which is bad. If you don't want to have a default param in `__init__`, then please remove `.get(...)` from here.

Reply:
Hm, I don't completely understand what you mean here =)